### PR TITLE
CI for contra-env-setup

### PR DESCRIPTION
**Pull request by dirgim**
CI for contra-env-setup implemented with a scripted Jenkins pipeline for basic CI with:
* GitHub ghprb trigger configured for webhooks
* Shared libraries: cico, ci-pipeline and local
* Uses the testing container to run the tests - just centos7 (contra-env-setup-test-c7) initially
* Archives generated pytest .log artifacts

The contra-env-setup-test-c7 container image is used for testing contra-env-setup on a centos7 based system.
The testing consists of a prepare-and-test.sh script which does the following:

- sets up the environment (prepares Ara, clones the repository and fetches the desired PR)
- runs the contra-env-setup playbook for the sample project (this project) with the included playbook hook which outputs some debug variables necessary for testing
- runs pytest on the test_contra_env_setup.py which tests the results of the playbook run
- saves the test logs

The test_contra_env_setup.py tests the following:    

- results of the playbook run (did the playbook run succesfuly, did any tasks run etc.)
- if the installed binaries are in place (minishift and oc)
- if the minishift profile is created and active
- if key openshift buildconfigs, builds, imagestreams and services exist
- if the jenkins master pod is running
- the route to the jenkins master instance and if the login page is active

### CODE REVIEW

**Comment by dirgim** [comment]
[test]

**Comment by dirgim** [comment]
[test]

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+    env.DOCKER_REPO_URL = env.DOCKER_REPO_URL ?: '172.30.254.79:5000'
+    env.OPENSHIFT_NAMESPACE = env.OPENSHIFT_NAMESPACE ?: 'continuous-infra'
+    env.OPENSHIFT_SERVICE_ACCOUNT = env.OPENSHIFT_SERVICE_ACCOUNT ?: 'jenkins'
+
+    // Execution ID for this run of the pipeline
+    executionID = UUID.randomUUID().toString()
+
+    // Pod name to use
+    podName = 'env-setup-' + executionID
```
Comment (arilivigni) [inline Jenkinsfile]: contra-env-setup- to be explicit

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                                            description: 'Pull Request Number'),
+                                    string(name: 'ghprbPullAuthorLogin',
+                                            defaultValue: '',
+                                            description: 'Pull Request Author username'),
+                                    string(name: 'SLAVE_TAG',
+                                            defaultValue: 'stable',
+                                            description: 'Tag for slave image'),
+                                    string(name: 'ENVSETUPTEST_TAG',
+                                            defaultValue: 'stable',
+                                            description: 'Tag for env-setup-test-c7 image'),
```
Comment (arilivigni) [inline Jenkinsfile]: contra-env-setup-

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+            containers: [
+                    // This adds the custom slave container to the pod. Must be first with name 'jnlp'
+                    containerTemplate(name: 'jnlp',
+                            image: DOCKER_REPO_URL + '/' + OPENSHIFT_NAMESPACE + '/jenkins-continuous-infra-slave:' + SLAVE_TAG,
+                            ttyEnabled: false,
+                            args: '${computer.jnlpmac} ${computer.name}',
+                            command: '',
+                            workingDir: '/workDir'),
+                    // This adds the env-setup-test-c7 container to the pod.
```
Comment (arilivigni) [inline Jenkinsfile]: contra-env-setup-

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+                    containerTemplate(name: 'env-setup-test-c7',
```
Comment (arilivigni) [inline Jenkinsfile]: contra-env-setup-

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                            alwaysPullImage: true,
+                            image: DOCKER_REPO_URL + '/' + OPENSHIFT_NAMESPACE + '/env-setup-test-c7:' + ENVSETUPTEST_TAG,
```
Comment (arilivigni) [inline Jenkinsfile]: contra-env-setup-

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+                                currentStage = "test-env-setup"
+                                stage(currentStage) {
+
+                                    envsetupUtils.timedPipelineStep(stepName: currentStage, debug: true) {
+                                        // Set stage specific vars
+                                        envsetupUtils.setStageEnvVars(currentStage)
+
+                                        // Run env-setup test
+                                        pipelineUtils.executeInContainer(currentStage, "env-setup-test-c7", "/home/prepare_and_test.sh")
```
Comment (arilivigni) [inline Jenkinsfile]: contra-env-setup-

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    libraries.each { name, repo ->
+        library identifier: "${name}@${repo[0]}",
+                retriever: modernSCM([$class: 'GitSCMSource',
+                                      remote: repo[1]])
+    }
+
+    library identifier: 'contra-env-setup@env-setup-ci',
+            retriever: modernSCM([$class: 'GitSCMSource', credentialsId: '',
+                                  id: '25651a23-bada-49f0-8b7a-2d735623d646',
+                                  remote: 'https://github.com/dirgim/contra-env-setup.git',
```
Comment (johnbieren) [inline Jenkinsfile]: probably want ghprb vars here

Comment (dirgim) [reply Jenkinsfile]: Agree, will fix that

**Comment by dirgim** [comment]
[test]

**Comment by dirgim** [comment]
[test]

File: test/config/s2i/contra-env-setup-test-c7/contra-env-setup-test-buildconfig-template.yml [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    - type: ConfigChange
+parameters:
+- description: Git repository with Dockerfile and slave entrypoint.
+  displayName: Repository URL
+  name: REPO_URL
+  value: https://github.com/CentOS-PaaS-SIG/contra-env-setup.git
+- description: The sub-directory inside the repository.
+  displayName: Context Directory
+  name: REPO_CONTEXTDIR
+  value: tests/config/Dockerfiles/contra-env-setup-test-c7
```
[the remark this answers is not in the document]
Comment (dirgim) [reply test/config/s2i/contra-env-setup-test-c7/contra-env-setup-test-buildconfig-template.yml]: Fixed.

**Pull request merged**